apple/swift-docc
# drop uncurated symbols from the navigator index

Bug/issue #, if applicable: rdar://125855186

## Summary

When building the navigator index, it first collects the nodes into a tree based on their topic group curation. Then, for any node that is not curated into another node, it inserts them all into the navigator root. However, this step can erroneously create a situation where symbols and articles can appear outside a rendered doc archive, depending on how the archive is being loaded. While this is useful and necessary for articles (if articles are not curated into a bundle and DocC is processing more than bundle, adding them outside the hierarchy is still useful), symbols should all already be curated into place via automatic curation. If automatic curation dropped a symbol (e.g. when an overloaded symbol is being condensed into an overload group), it should not suddenly reappear into the root of the navigator, outside of any documented module.

This PR adds a filter to the navigator index to prevent this root-insertion for symbol nodes.

## Dependencies

None

## Testing

Unfortunately, the nodes inserted at the root of the navigator tree only appear in the in-memory and LMDB representations of the index, so the bug doesn't manifest when building and previewing in the browser with Swift-DocC-Render. However, Xcode uses the LMDB index to show its navigator, so we can perform the test there.

Create an Xcode project (not a Swift package - we need to edit build settings) that includes the following Swift code:

```swift
public class MyClass {
    public func myFunc(param: String) {}
    public func myFunc(param: Int) {}
}
```

Steps:
1. In the project build settings, set `DOCC_EXEC` to your local build of Swift-DocC. (This needs to be set manually, via the plus-shaped button's "Add user-defined setting" option in the viewer.)
2. In the project build settings, add `--enable-experimental-overloaded-symbol-presentation` to the "Other DocC Flags" build setting (a.k.a. `OTHER_DOCC_FLAGS`).
3. Build documentation for the target that includes the above code.
4. Ensure that the individual `MyClass.myFunc(param:)` methods do not appear outside the target's documentation bundle:

<details>
<summary>Before</summary>

<img width="233" alt="framework documentation sidebar for 'SwiftDemo', showing excess methods after the framework" src="https://github.com/apple/swift-docc/assets/5217170/8fb7fb03-6594-4df6-a44b-d44295ca68d4">

</details>

<details>
<summary>After</summary>

<img width="229" alt="framework documentation sidebar for 'SwiftDemo', showing no excess methods" src="https://github.com/apple/swift-docc/assets/5217170/43f91bfc-e9cf-4d57-a72a-622047402e6d">

</details>

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ n/a ] Updated documentation if necessary
